src/server/generic_family.cc

```diff
@@ -243,161 +253,189 @@ OpResult<RestoreArgs> RestoreArgs::TryFrom(const CmdArgList& args) {
+    return;
+  }
+
+  if (dest_found_ && destination_should_not_exist) {
+    status_ = OpStatus::KEY_EXISTS;
+    transaction_->Conclude();
+    return;
+  }
+
+  FinalizeRename();
+}
+
+void Renamer::FetchData() {
   auto cb = [this](Transaction* t, EngineShard* shard) {
     auto args = t->GetShardArgs(shard->shard_id());
     DCHECK_EQ(1u, args.Size());
 
-    FindResult* res = (shard->shard_id() == src_sid_) ? &src_res_ : &dest_res_;
+    const ShardId shard_id = shard->shard_id();
 
-    res->key = args.Front();
-    auto& db_slice = EngineShard::tlocal()->db_slice();
-    auto [it, exp_it] = db_slice.FindReadOnly(t->GetDbContext(), res->key);
+    if (shard_id == src_sid_) {
+      SerializeSrc(t, shard);
+    }
 
-    res->found = IsValid(it);
-    if (res->found) {
-      res->ref_val = it->second.AsRef();
-      res->expire_ts = db_slice.ExpireTime(exp_it);
-      res->sticky = it->first.IsSticky();
+    if (shard_id == dest_sid_) {
+      dest_found_ = KeyExists(t, shard, dest_key_);
     }
+
     return OpStatus::OK;
   };
 
-  t->Execute(std::move(cb), false);
-};
+  transaction_->Execute(std::move(cb), false);
+}
 
-void Renamer::Finalize(Transaction* t, bool skip_exist_dest) {
-  if (!src_res_.found) {
-    status_ = OpStatus::KEY_NOTFOUND;
-    t->Conclude();
-    return;
-  }
+void Renamer::FinalizeRename() {
+  auto cb = [this](Transaction* t, EngineShard* shard) {
+    const ShardId shard_id = shard->shard_id();
 
-  if (dest_res_.found && skip_exist_dest) {
-    status_ = OpStatus::KEY_EXISTS;
-    t->Conclude();
+    if (shard_id == src_sid_) {
+      return DelSrc(t, shard);
+    }
+
+    if (shard_id == dest_sid_) {
+      return DeserializeDest(t, shard);
+    }
+
+    return OpStatus::OK;
+  };
+
+  transaction_->Execute(std::move(cb), true);
+}
+
+bool Renamer::KeyExists(Transaction* t, EngineShard* shard, std::string_view key) const {
+  auto& db_slice = shard->db_slice();
+  auto it = db_slice.FindReadOnly(t->GetDbContext(), key).it;
+  return IsValid(it);
+}
+
+void Renamer::SerializeSrc(Transaction* t, EngineShard* shard) {
+  auto& db_slice = shard->db_slice();
+  auto [it, exp_it] = db_slice.FindReadOnly(t->GetDbContext(), src_key_);
+
+  src_found_ = IsValid(it);
+  if (!src_found_) {
     return;
   }
 
-  DCHECK(src_res_.ref_val.IsRef());
+  DVLOG(1) << "Rename: key '" << src_key_ << "' successfully found, going to dump it";
 
-  // Src key exist and we need to override the destination.
-  // Alternatively, we could apply an optimistic algorithm and move src at Find step.
-  // We would need to restore the state in case of cleanups.
-  t->Execute([&](Transaction* t, EngineShard* shard) { return MoveSrc(t, shard); }, false);
-  t->Execute([&](Transaction* t, EngineShard* shard) { return UpdateDest(t, shard); }, true);
+  io::StringSink sink;
+  SerializerBase::DumpObject(it->second, &sink);
+  serialized_value_ = {std::move(sink).str(), db_slice.ExpireTime(exp_it), it->first.IsSticky()};
```

Comment:
you dont need std::move on sink

Reply:
maybe I'm missing something, but without `std::move`, there will be a copy here. With `std::move` here, [another str() function](https://github.com/romange/helio/pull/201/commits/de9d1a7e88b776d4790342a4e0ff5ff9eb63a32d) is called.